<a href="https://colab.research.google.com/github/jhj030722/toy_project_2024/blob/main/%ED%86%A0%EC%9D%B4%ED%94%84%EB%A1%9C%EC%A0%9D%ED%8A%B8__EDA_%EC%A3%BC%ED%98%84.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 분석의 과정 로드맵

**a. (겉으로 드러난 현상)**

- 결측치가 매우 매우 많다 ㅠ.ㅠ
- 결측치 핸들링이 중요할 거같음
- 데이터 수가 많지 않다..(5천개도 안됨)

**b. 목적 및 문제를 정의**

- 분석 목적: 술은 몸에 안좋다.. *최소 얼마나 술을 끊어야지 유의미한 건강 변화가 있을까?*

- 분석할 문제:
  - **술을 얼마나 자주 마셔야 건강에 나쁠까? : 주간/월간 음주 횟수에 따른 성인병 위험도 변화 분석**

  - **특정 술 종류와 특정 성인병(예: 맥주와 고지혈증) 간의 연관성 파악**

  - **음주 중단 후 성인병 지표(혈압, 혈당 등)의 변화 추적**

  - **사회경제적 지위(교육 수준?)에 따른 음주 유형과 음주 양의 차이**

- 모델 구축?

**c. 지표(데이터)를 결정**

- 위 데이터에서 음주 습관과 관련된 지표들

- 혈압, 인슐린 수치, 혈당, 콜레스테롤 수치 등의 지표

**d. 현재 상태를 파악**

<결측치 파악>
- 결측치가 매우 매우 많다 ㅠ.ㅠ
- 결측치 핸들링이 중요할 거같음
- 데이터 수가 많지 않다..(5천개도 안됨)

<EDA 과정 (시각화 및 변수파악)>

시각화 - 이상치와 특이점 찾기

<가설검정 - 가설설정 및 검증>

가설(예시):


**e. 평가**

비교결과 분석 및 결론도출
의미있는 인사이트 제시

**f. 원인을 분석**

**g. 해결방안을 모색**

해결방안 제안

추가과제 탐색

## 데이터 로드

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
df1 = pd.read_csv('/content/drive/MyDrive/DART-B/toy 24-2/follow_01_data.csv')
df2 = pd.read_csv('/content/drive/MyDrive/DART-B/toy 24-2/follow_02_data.csv')
df3 = pd.read_csv('/content/drive/MyDrive/DART-B/toy 24-2/follow_03_data.csv')
df4 = pd.read_csv('/content/drive/MyDrive/DART-B/toy 24-2/follow_04_data.csv')
df5 = pd.read_csv('/content/drive/MyDrive/DART-B/toy 24-2/follow_05_data.csv')

In [ ]:
df = pd.read_excel('/content/drive/MyDrive/DART-B/toy 24-2/Sorted_Merged_Data.xlsx')

In [ ]:
df

T00_ID T00_DATA_CLASS  T00_SEX  T01_EDU  T01_MARRY  T02_MARRY  \
0    K_FOLLOW_0001            F05        1        1          2      66666   
1    K_FOLLOW_0002            F19        1        3          2          2   
2    K_FOLLOW_0003            F05        1        2          2      66666   
3    K_FOLLOW_0004            F05        2        2          2      66666   
4    K_FOLLOW_0005            F19        1        3          2          2   
..             ...            ...      ...      ...        ...        ...   
832  K_FOLLOW_0996            F05        1        1          2      66666   
833  K_FOLLOW_0997            F18        1        3          2          2   
834  K_FOLLOW_0998            F19        1        5          2          2   
835  K_FOLLOW_0999            F11        2        1          2          2   
836  K_FOLLOW_1000            F19        2        2          2          2   

     T03_MARRY  T04_MARRY  T01_AGE  T02_AGE  ...  T03_WEIGHT  T04_WEIGHT  \
0            2          2       56       58  ...          51          52   
1            2          2       40       41  ...          90          93   
2            2          2       52       54  ...          67          67   
3            2          3       60       62  ...          71          70   
4            2          2       49       51  ...          65          67   
..         ...        ...      ...      ...  ...         ...         ...   
832          2          2       64       66  ...          55          59   
833          2          2       63       65  ...          73          75   
834          2          2       41       43  ...          76          75   
835          2          2       69       71  ...          47          45   
836          2          2       45       47  ...          69          71   

     T01_WINEAM  T02_WINEAM  T03_WINEAM  T04_WINEAM  T01_WINEFQ  T02_WINEFQ  \
0       77777.0     77777.0     77777.0     77777.0           0           0   
1       77777.0     77777.0     77777.0     77777.0           0           0   
2       77777.0     77777.0     77777.0     77777.0           0           0   
3       77777.0     77777.0     77777.0     77777.0           0           0   
4       77777.0     77777.0     77777.0     77777.0           0           0   
..          ...         ...         ...         ...         ...         ...   
832     77777.0     77777.0     77777.0     77777.0           0           0   
833     77777.0     77777.0     77777.0     77777.0           0           0   
834     77777.0     77777.0     77777.0     77777.0           0           0   
835     77777.0     77777.0     77777.0     77777.0           0           0   
836     77777.0     77777.0     77777.0     77777.0           0           0   

     T03_WINEFQ  T04_WINEFQ  
0             0           0  
1             0           0  
2             0           0  
3             0           0  
4             0           0  
..          ...         ...  
832           0           0  
833           0           0  
834           0           0  
835           0           0  
836           0           0  

[837 rows x 223 columns]

# 데이터 설명

데이터셋의 각 열은 다음을 나타냅니다:

- 개인 식별자 => 삭제

- 데이터 단위 : T00_DATA_CLASS

- 설문 날짜 (YYYYMM) : TOO_EDATE

- 성별 1 = 남성, 2 = 여성 : TOO_AGE

- 나이 (세)

- 교육 수준 1 = 초등학교 이하, 2 = 중학교, 3 = 고등학교, 4 = 학사 학위, 5 = 석사 학위 이상

- 결혼 상태 1 = 미혼, 2 = 기혼, 3 = 사별, 4 = 별거, 5 = 이혼, 6 = 기타

- 음주 여부 1 = 비음주자, 2 = 과거 음주자, 3 = 현재 음주자

- 총 음주 기간 1 = 5년 이하, 2 = 6-10년, 3 = 11-20년, 4 = 21년 이상 => 이 컬럼 왜 없지??

- 막걸리 음주 빈도 0 = 없음, 1 = 한 달에 한 번, 2 = 한 달에 2-3번, 3 = 일주일에 한 번, 4 = 일주일에 2-3번, 5 = 일주일에 4-6번, 6 = 매일

- 한 번에 마시는 막걸리 양 () 잔

- 사케 음주 빈도 0 = 없음, 1 = 한 달에 한 번, 2 = 한 달에 2-3번, 3 = 일주일에 한 번, 4 = 일주일에 2-3번, 5 = 일주일에 4-6번, 6 = 매일

- 한 번에 마시는 사케 양 () 잔

- 와인 음주 빈도 0 = 없음, 1 = 한 달에 한 번, 2 = 한 달에 2-3번, 3 = 일주일에 한 번, 4 = 일주일에 2-3번, 5 = 일주일에 4-6번, 6 = 매일

- 한 번에 마시는 와인 양 () 잔

- 소주 음주 빈도 0 = 없음, 1 = 한 달에 한 번, 2 = 한 달에 2-3번, 3 = 일주일에 한 번, 4 = 일주일에 2-3번, 5 = 일주일에 4-6번, 6 = 매일

- 한 번에 마시는 소주 양 () 잔

- 맥주 음주 빈도 0 = 없음, 1 = 한 달에 한 번, 2 = 한 달에 2-3번, 3 = 일주일에 한 번, 4 = 일주일에 2-3번, 5 = 일주일에 4-6번, 6 = 매일

한 번에 마시는 맥주 양 () 병

증류주 음주 빈도 0 = 없음, 1 = 한 달에 한 번, 2 = 한 달에 2-3번, 3 = 일주일에 한 번, 4 = 일주일에 2-3번, 5 = 일주일에 4-6번, 6 = 매일

- 한 번에 마시는 증류주 양 () 샷

- 흡연 상태 1 = 비흡연자, 2 = 과거 흡연자, 3 = 현재 흡연자

- 흡연 시작 나이 () 세

- 총 흡연 기간 () 년

- 하루 흡연량 () 갑

- 간접흡연 여부 1 = 없음, 2 = 있음

- 고강도 운동 여부 1 = 없음, 2 = 있음

- 고혈압 병력 1 = 없음, 2 = 있음

- 고혈압 진단 나이 () 세

- 당뇨병 진단 여부 1 = 없음, 2 = 있음

- 당뇨병 진단 나이 () 세

- 고지혈증 진단 여부 1 = 없음, 2 = 있음

- 고지혈증 진단 나이 () 세

- 부모의 고혈압 병력 (아버지) 1 = 없음, 2 = 있음

- 아버지의 고혈압 진단 나이 () 세

- 부모의 고혈압 병력 (어머니) 1 = 없음, 2 = 있음

- 어머니의 고혈압 진단 나이 () 세

- 부모의 당뇨병 병력 (아버지) 1 = 없음, 2 = 있음

- 아버지의 당뇨병 진단 나이 () 세

- 부모의 당뇨병 병력 (어머니) 1 = 없음, 2 = 있음

- 어머니의 당뇨병 진단 나이 () 세

- 초경 나이 () 세

- 임신 경험 1 = 없음, 2 = 있음

- 첫 임신 나이 () 세

- 출산 여부 1 = 없음, 2 = 있음

- 첫 출산 나이 () 세

- 폐경 여부 1 = 없음, 2 = 있음

- 폐경 나이 () 세

- 신장 () cm

- 체중 () kg

- 평균 허리둘레 () cm

- 평균 엉덩이둘레 () cm

- 평균 심박수 () 회/분

- 수축기 혈압 () mmHg

- 이완기 혈압 () mmHg

- HbA1C ()%

- 공복 혈당 () mg/dL

- 크레아티닌 () mg/dL

- AST () IU/L

- ALT () IU/L

- 총 콜레스테롤 () mg/dL

- HDL 콜레스테롤 () mg/dL

- 중성지방 () mg/dL

- 공복 인슐린 () µIU/mL

- 참고

[코드 66666: 조사되지 않음; 77777: 해당 없음; 99999: 알 수 없음/무응답/측정되지 않음]

# (1) 데이터 전처리

In [ ]:
# 66666 (조사되지 않음), 77777 (해당 없음), 99999 (알 수 없음/응답 없음/측정되지 않음)와 같은 특수 코드를 결측치로 변환

df.replace([66666, 77777, 99999], np.nan, inplace=True)

In [ ]:
# 한글 깨짐 방지

!sudo apt-get install -y fonts-nanum
!sudo fc-cache -fv
!rm ~/.cache/matplotlib -rf

Reading package lists... Done

Reading state information... Done
fonts-nanum is already the newest version (20200506-1).
0 upgraded, 0 newly installed, 0 to remove and 49 not upgraded.
/usr/share/fonts: caching, new cache contents: 0 fonts, 1 dirs
/usr/share/fonts/truetype: caching, new cache contents: 0 fonts, 3 dirs
/usr/share/fonts/truetype/humor-sans: caching, new cache contents: 1 fonts, 0 dirs
/usr/share/fonts/truetype/liberation: caching, new cache contents: 16 fonts, 0 dirs
/usr/share/fonts/truetype/nanum: caching, new cache contents: 12 fonts, 0 dirs
/usr/local/share/fonts: caching, new cache contents: 0 fonts, 0 dirs
/root/.local/share/fonts: skipping, no such directory
/root/.fonts: skipping, no such directory
/usr/share/fonts/truetype: skipping, looped directory detected
/usr/share/fonts/truetype/humor-sans: skipping, looped directory detected
/usr/share/fonts/truetype/liberation: skipping, looped directory detected
/usr/share/fonts/truetype/nanum: skipping, looped director

## 한글로 변환

In [ ]:
# Create a function to dynamically map the column names based on the prefixes (T01, T02, T03, T04)
def generate_timepoint_column_mapping(base_mapping):
    timepoints = {
        'T01': '1차',
        'T02': '2차',
        'T03': '3차',
        'T04': '4차'
    }
    full_mapping = {}

    for time_prefix, time_suffix in timepoints.items():
        for col, trans in base_mapping.items():
            full_mapping[f'{time_prefix}_{col}'] = f'{trans}_{time_suffix}'

    return full_mapping

# Base mapping provided by the user
base_mapping = {
    'DATA_CLASS': '데이터_단위',
    'EDATE': '설문_날짜',
    'AGE': '나이',
    'MARRY': '결혼_상태',
    'DRINK': '음주_여부',
    'TAKFQ': '막걸리_음주_빈도',
    'TAKAM': '막걸리_양',
    'RICEFQ': '사케_음주_빈도',
    'RICEAM': '사케_양',
    'WINEFQ': '와인_음주_빈도',
    'WINEAM': '와인_양',
    'SOJUFQ': '소주_음주_빈도',
    'SOJUAM': '소주_양',
    'BEERFQ': '맥주_음주_빈도',
    'BEERAM': '맥주_양',
    'HLIQFQ': '증류주_음주_빈도',
    'HLIQAM': '증류주_양',
    'SMOKE': '흡연_상태',
    'SMAG': '흡연_시작_나이',
    'SMDU': '총_흡연_기간',
    'SMAM': '하루_흡연량',
    'PSM': '간접흡연_여부',
    'EXER': '고강도_운동_여부',
    'HTN': '고혈압_병력',
    'HTNAG': '고혈압_진단_나이',
    'DM': '당뇨병_진단_여부',
    'DMAG': '당뇨병_진단_나이',
    'LIP': '고지혈증_진단_여부',
    'LIPAG': '고지혈증_진단_나이',
    'FMFHT': '부모_고혈압_병력_아버지',
    'FMFHTAG': '아버지_고혈압_진단_나이',
    'FMMHT': '부모_고혈압_병력_어머니',
    'FMMHTAG': '어머니_고혈압_진단_나이',
    'FMFDM': '부모_당뇨병_병력_아버지',
    'FMFDMAG': '아버지_당뇨병_진단_나이',
    'FMMDM': '부모_당뇨병_병력_어머니',
    'FMMDMAG': '어머니_당뇨병_진단_나이',
    'MNSAG': '초경_나이',
    'PREG': '임신_경험',
    'FPREGAG': '첫_임신_나이',
    'CHILD': '출산_여부',
    'FLABAG': '첫_출산_나이',
    'PMYN_C': '폐경_여부',
    'PMAG_C': '폐경_나이',
    'HEIGHT': '신장',
    'WEIGHT': '체중',
    'WAIST': '평균_허리둘레',
    'HIP': '평균_엉덩이둘레',
    'PULSE': '평균_심박수',
    'SBP': '수축기_혈압',
    'DBP': '이완기_혈압',
    'HBA1C': 'HbA1C',
    'GLU0': '공복_혈당',
    'CREATININE': '크레아티닌',
    'AST': 'AST',
    'ALT': 'ALT',
    'TCHL': '총_콜레스테롤',
    'HDL': 'HDL_콜레스테롤',
    'TG': '중성지방',
    'INS0': '공복_인슐린'
}

# Generate the full mapping for all time points (T01, T02, T03, T04)
full_timepoint_mapping = generate_timepoint_column_mapping(base_mapping)

# Now apply this full mapping to the dataset
df_renamed_with_timepoints = df.rename(columns=full_timepoint_mapping)

# Display the first few columns to confirm renaming
df_renamed_with_timepoints.columns[:20]  # Display only a sample of the columns to check


Index(['T00_ID', 'T00_DATA_CLASS', 'T00_SEX', 'T01_EDU', '결혼_상태_1차',
       '결혼_상태_2차', '결혼_상태_3차', '결혼_상태_4차', '나이_1차', '나이_2차', '나이_3차', '나이_4차',
       '설문_날짜_1차', '설문_날짜_2차', '설문_날짜_3차', '설문_날짜_4차', 'ALT_1차', 'ALT_2차',
       'ALT_3차', 'ALT_4차'],
      dtype='object')

In [ ]:
df_renamed_with_timepoints

# 엑셀 파일로 다운로드
df_renamed_with_timepoints.to_excel('renamed_data_with_timepoints.xlsx', index=False)

In [ ]:
merge_df = pd.read_excel('/content/renamed_data_with_timepoints.xlsx')

merge_df.head()

T00_ID T00_DATA_CLASS  T00_SEX  T01_EDU  결혼_상태_1차  결혼_상태_2차  \
0  K_FOLLOW_0001            F05        1      1.0       2.0       NaN   
1  K_FOLLOW_0002            F19        1      3.0       2.0       2.0   
2  K_FOLLOW_0003            F05        1      2.0       2.0       NaN   
3  K_FOLLOW_0004            F05        2      2.0       2.0       NaN   
4  K_FOLLOW_0005            F19        1      3.0       2.0       2.0   

   결혼_상태_3차  결혼_상태_4차  나이_1차  나이_2차  ...  체중_3차  체중_4차  와인_양_1차  와인_양_2차  \
0       2.0       2.0     56     58  ...   51.0     52      NaN      NaN   
1       2.0       2.0     40     41  ...   90.0     93      NaN      NaN   
2       2.0       2.0     52     54  ...   67.0     67      NaN      NaN   
3       2.0       3.0     60     62  ...   71.0     70      NaN      NaN   
4       2.0       2.0     49     51  ...   65.0     67      NaN      NaN   

   와인_양_3차  와인_양_4차  와인_음주_빈도_1차  와인_음주_빈도_2차  와인_음주_빈도_3차  와인_음주_빈도_4차  
0      NaN      NaN          0.0          0.0          0.0          0.0  
1      NaN      NaN          0.0          0.0          0.0          0.0  
2      NaN      NaN          0.0          0.0          0.0          0.0  
3      NaN      NaN          0.0          0.0          0.0          0.0  
4      NaN      NaN          0.0          0.0          0.0          0.0  

[5 rows x 223 columns]

## 결측치

In [ ]:
import os
from os.path import join
import copy
import warnings
warnings.filterwarnings('ignore')


#Plots
import missingno as msno
from sklearn.pipeline import make_pipeline
import plotly.graph_objs as go
import plotly.offline as py

def missing_plot(merge_df, name):
    null_feat = pd.DataFrame(len(merge_df) - merge_df.isnull().sum(), columns=['Count'])
    percentage_null = pd.DataFrame((merge_df.isnull().sum() / len(merge_df)) * 100, columns=['Percentage'])
    percentage_null = percentage_null.round(2)

    trace = go.Bar(x=null_feat.index, y=null_feat['Count'], opacity=0.8, text=percentage_null['Percentage'],
                   textposition='auto', marker=dict(color='#7EC0EE', line=dict(color='#000000', width=1.5)))

    layout = dict(title=name)

    fig = dict(data=[trace], layout=layout)
    py.iplot(fig)

# 전체 데이터프레임의 결측치 시각화
missing_plot(merge_df, 'Missing Values (count & %)')


# (2) 목적 및 문제 정의

✅   성인병(당뇨 고지혈증 고혈압) 발병에는 음주 여부보다 유전적 요인이 영향을 미친다.

In [ ]:
# sm module 불러오기
import statsmodels.api as sm

##결측치 채우기

데이터 결측치 채우기

선형 보간법 (Linear Interpolation)
  - 설명: 시간 순서에 따라 결측값을 앞뒤 값의 선형 보간을 통해 채우는 방법입니다. 시계열 데이터에 유용합니다.
  - 장점: 시간에 따른 변동을 반영할 수 있습니다.
  - 단점: 데이터가 시간 순서에 맞지 않거나 비선형일 경우 왜곡이 발생할 수 있습니다

In [ ]:
# 각 차수별로 필요한 변수 리스트
variables_to_filter = [ 'T00_ID',
    '당뇨병_진단_여부_1차', '당뇨병_진단_여부_2차', '당뇨병_진단_여부_3차', '당뇨병_진단_여부_4차',
    '고지혈증_진단_여부_1차', '고지혈증_진단_여부_2차', '고지혈증_진단_여부_3차', '고지혈증_진단_여부_4차',
    '고혈압_병력_1차', '고혈압_병력_2차', '고혈압_병력_3차', '고혈압_병력_4차',
    '음주_여부_1차', '음주_여부_2차', '음주_여부_3차', '음주_여부_4차',
    '부모_고혈압_병력_아버지_1차', '부모_고혈압_병력_아버지_2차', '부모_고혈압_병력_아버지_3차', '부모_고혈압_병력_아버지_4차',
    '부모_고혈압_병력_어머니_1차', '부모_고혈압_병력_어머니_2차', '부모_고혈압_병력_어머니_3차', '부모_고혈압_병력_어머니_4차',
    '부모_당뇨병_병력_아버지_1차', '부모_당뇨병_병력_아버지_2차', '부모_당뇨병_병력_아버지_3차', '부모_당뇨병_병력_아버지_4차',
    '부모_당뇨병_병력_어머니_1차', '부모_당뇨병_병력_어머니_2차', '부모_당뇨병_병력_어머니_3차', '부모_당뇨병_병력_어머니_4차'
]

# merge_df에서 필요한 변수만 필터링하여 새로운 데이터프레임 생성 (결측치는 최빈값으로 채우기)

filled_merge_df = merge_df[variables_to_filter].fillna(merge_df[variables_to_filter].mode().iloc[0])

filled_merge_df.isnull().sum()

T00_ID              0
당뇨병_진단_여부_1차        0
당뇨병_진단_여부_2차        0
당뇨병_진단_여부_3차        0
당뇨병_진단_여부_4차        0
고지혈증_진단_여부_1차       0
고지혈증_진단_여부_2차       0
고지혈증_진단_여부_3차       0
고지혈증_진단_여부_4차       0
고혈압_병력_1차           0
고혈압_병력_2차           0
고혈압_병력_3차           0
고혈압_병력_4차           0
음주_여부_1차            0
음주_여부_2차            0
음주_여부_3차            0
음주_여부_4차            0
부모_고혈압_병력_아버지_1차    0
부모_고혈압_병력_아버지_2차    0
부모_고혈압_병력_아버지_3차    0
부모_고혈압_병력_아버지_4차    0
부모_고혈압_병력_어머니_1차    0
부모_고혈압_병력_어머니_2차    0
부모_고혈압_병력_어머니_3차    0
부모_고혈압_병력_어머니_4차    0
부모_당뇨병_병력_아버지_1차    0
부모_당뇨병_병력_아버지_2차    0
부모_당뇨병_병력_아버지_3차    0
부모_당뇨병_병력_아버지_4차    0
부모_당뇨병_병력_어머니_1차    0
부모_당뇨병_병력_어머니_2차    0
부모_당뇨병_병력_어머니_3차    0
부모_당뇨병_병력_어머니_4차    0
dtype: int64

In [ ]:
filled_merge_df.isnull().sum()

T00_ID              0
당뇨병_진단_여부_1차        0
당뇨병_진단_여부_2차        0
당뇨병_진단_여부_3차        0
당뇨병_진단_여부_4차        0
고지혈증_진단_여부_1차       0
고지혈증_진단_여부_2차       0
고지혈증_진단_여부_3차       0
고지혈증_진단_여부_4차       0
고혈압_병력_1차           0
고혈압_병력_2차           0
고혈압_병력_3차           0
고혈압_병력_4차           0
음주_여부_1차            0
음주_여부_2차            0
음주_여부_3차            0
음주_여부_4차            0
부모_고혈압_병력_아버지_1차    0
부모_고혈압_병력_아버지_2차    0
부모_고혈압_병력_아버지_3차    0
부모_고혈압_병력_아버지_4차    0
부모_고혈압_병력_어머니_1차    0
부모_고혈압_병력_어머니_2차    0
부모_고혈압_병력_어머니_3차    0
부모_고혈압_병력_어머니_4차    0
부모_당뇨병_병력_아버지_1차    0
부모_당뇨병_병력_아버지_2차    0
부모_당뇨병_병력_아버지_3차    0
부모_당뇨병_병력_아버지_4차    0
부모_당뇨병_병력_어머니_1차    0
부모_당뇨병_병력_어머니_2차    0
부모_당뇨병_병력_어머니_3차    0
부모_당뇨병_병력_어머니_4차    0
dtype: int64



---



## 가설 전) 음주 여부와 가족력 변수의 독립성 검정

In [ ]:
# 교차표 작성: 부모의 병력과 음주 여부 간의 교차표
# 부모의 병력: 아버지와 어머니 중 하나라도 병력이 있는 경우를 1, 없는 경우를 0으로 표시
filled_merge_df['부모_당뇨병_병력_종합'] = ((filled_merge_df['아버지_당뇨병_병력_종합'] == 1) |
                                      (filled_merge_df['어머니_당뇨병_병력_종합'] == 1)).astype(int)

# 음주 여부와 부모의 병력에 대한 교차표
crosstab = pd.crosstab(filled_merge_df['꾸준히_비음주자'], filled_merge_df['부모_당뇨병_병력_종합'])

# 카이제곱 독립성 검정 수행
chi2, p, dof, expected = stats.chi2_contingency(crosstab)

# 결과 출력
print(f"Chi-squared Test Statistic: {chi2}")
print(f"p-value: {p}")
print(f"Degrees of Freedom: {dof}")
print(f"Expected Frequencies:\n{expected}")


Chi-squared Test Statistic: 0.9806884990596068
p-value: 0.32202888593595685
Degrees of Freedom: 1
Expected Frequencies:
[[471.68458781  53.31541219]
 [280.31541219  31.68458781]]


- h0: 음주여부와 가족력(당뇨병)은 통계적으로 독립이다.
- h1: 음주여부와 가족력(당뇨병)은 통계적으로 연관이 있다.

- P-value가 0.35로 0.05보다 크기 때문에 귀무가설을 기각할 수 없다
- **따라서 '음주 여부'와 '가족력(당뇨병)'은 통계적으로 독립적이다**.

In [ ]:
# 고혈압
filled_merge_df['부모_고혈압_병력_종합'] = ((filled_merge_df['아버지_고혈압_병력_종합'] == 1) |
                                      (filled_merge_df['어머니_고혈압_병력_종합'] == 1)).astype(int)

# 음주 여부와 부모의 병력에 대한 교차표
crosstab = pd.crosstab(filled_merge_df['꾸준히_비음주자'], filled_merge_df['부모_고혈압_병력_종합'])

# 카이제곱 독립성 검정 수행
chi2, p, dof, expected = stats.chi2_contingency(crosstab)

# 결과 출력
print(f"Chi-squared Test Statistic: {chi2}")
print(f"p-value: {p}")
print(f"Degrees of Freedom: {dof}")
print(f"Expected Frequencies:\n{expected}")


Chi-squared Test Statistic: 2.4170657200295267
p-value: 0.12001955536491325
Degrees of Freedom: 1
Expected Frequencies:
[[422.13261649 102.86738351]
 [250.86738351  61.13261649]]


- h0: 음주여부와 가족력(당뇨병)은 통계적으로 독립이다.
- h1: 음주여부와 가족력(당뇨병)은 통계적으로 연관이 있다.

P-value가 0.12로 0.05보다 크기 때문에 귀무가설을 기각할 수 없다

**따라서 '음주 여부'와 '가족력(고혈압)'은 통계적으로 독립적이다.**

##가설1) 성인병 발병에 음주여부가 영향을 미친다.

#### 당뇨병

In [ ]:
# 1차~4차 중 한번이라도 당뇨병 발병이 있으면 1, 그렇지 않으면 0
filled_merge_df['당뇨병_진단_여부_종합'] = (filled_merge_df['당뇨병_진단_여부_1차'] == 2) | \
                                 (filled_merge_df['당뇨병_진단_여부_2차'] == 2) | \
                                 (filled_merge_df['당뇨병_진단_여부_3차'] == 2) | \
                                 (filled_merge_df['당뇨병_진단_여부_4차'] == 2)

# 당뇨병 발병 여부를 0(발병 안함), 1(발병 있음)으로 변환
filled_merge_df['당뇨병_진단_여부_종합'] = filled_merge_df['당뇨병_진단_여부_종합'].astype(int)
filled_merge_df['당뇨병_진단_여부_종합'].value_counts()

당뇨병_진단_여부_종합
0    740
1     97
Name: count, dtype: int64

귀무가설(H0): 음주 여부와 당뇨병 발병은 서로 독립적이다.

대립가설(H1): 음주 여부와 당뇨병 발병은 독립적이지 않다 (연관성이 있다).

In [ ]:
# 꾸준히 비음주자인 사람 필터링 (음주 여부가 1(비음주자)인 경우)
filled_merge_df['꾸준히_비음주자'] = (filled_merge_df['음주_여부_1차'] == 1) & \
                           (filled_merge_df['음주_여부_2차'] == 1) & \
                           (filled_merge_df['음주_여부_3차'] == 1) & \
                           (filled_merge_df['음주_여부_4차'] == 1)

# 꾸준히 비음주자인 사람은 1, 그렇지 않은 사람은 0으로 변환
filled_merge_df['꾸준히_비음주자'] = filled_merge_df['꾸준히_비음주자'].astype(int)


In [ ]:
import pandas as pd
import scipy.stats as stats

# 꾸준히 비음주자인 사람과 아닌 사람 간의 당뇨병 진단 여부에 대한 교차표 작성
crosstab = pd.crosstab(filled_merge_df['꾸준히_비음주자'], filled_merge_df['당뇨병_진단_여부_종합'])

# 카이제곱 검정 수행
chi2, p, dof, expected = stats.chi2_contingency(crosstab)

# 결과 출력
print(f"카이제곱 통계량: {chi2}")
print(f"p-value: {p}")


카이제곱 통계량: 6.415947713595262
p-value: 0.011309995110545132


- p-값이 0.05보다 작으면, ***꾸준히 비음주자인 사람과 그렇지 않은 사람 간에 당뇨병 발병 여부가 통계적으로 유의미한 차이가 있다***고 해석할 수 있습니다.
- p-값이 0.05보다 크면, 두 그룹 간의 당뇨병 발병 여부에 유의미한 차이가 없다고 해석할 수 있습니다.

In [ ]:
crosstab

# 꾸준히 비음주자이며 당뇨병 발병 없는 사람: 1-0 이므로 261명
# 꾸준히 비음주자이며 당뇨병 발병 있는사람: 1-1이므로 48명
# 음주 경험 있으며 당뇨병 발병 없는 사람: 0-0이므로 479명
# 음주 경험 있으며 당뇨병 발병 있는 사람 0-1이므로 49명

당뇨병_진단_여부_종합    0   1
꾸준히_비음주자             
0             476  49
1             264  48

- 당뇨병 발병 경험이 있는 사람은 1, 한번도 없는 사람은 0
- 꾸준히 비음주자인 사람은 1, 음주 경험이 있는 사람은 0

#### 고지혈증

In [ ]:
# 1차~4차 중 한번이라도 고지혈증 발병이 있으면 1, 그렇지 않으면 0
filled_merge_df['고지혈증_진단_여부_종합'] = (filled_merge_df['고지혈증_진단_여부_1차'] == 2) | \
                                 (filled_merge_df['고지혈증_진단_여부_2차'] == 2) | \
                                 (filled_merge_df['고지혈증_진단_여부_3차'] == 2) | \
                                 (filled_merge_df['고지혈증_진단_여부_4차'] == 2)

# 고지혈증 발병 여부를 0(발병 안함), 1(발병 있음)으로 변환
filled_merge_df['고지혈증_진단_여부_종합'] = filled_merge_df['고지혈증_진단_여부_종합'].astype(int)
filled_merge_df['고지혈증_진단_여부_종합'].value_counts()

고지혈증_진단_여부_종합
0    720
1    117
Name: count, dtype: int64

In [ ]:
import pandas as pd
import scipy.stats as stats

# 꾸준히 비음주자인 사람과 아닌 사람 간의 당뇨병 진단 여부에 대한 교차표 작성
crosstab2 = pd.crosstab(filled_merge_df['꾸준히_비음주자'], filled_merge_df['고지혈증_진단_여부_종합'])

# 카이제곱 검정 수행
chi2, p, dof, expected = stats.chi2_contingency(crosstab2)

# 결과 출력
print(f"카이제곱 통계량: {chi2}")
print(f"p-value: {p}")


카이제곱 통계량: 6.004760843723586
p-value: 0.014267327756523744


- p-값이 0.05보다 작으면, ***꾸준히 비음주자인 사람과 그렇지 않은 사람 간에 고지혈증 발병 여부가 통계적으로 유의미한 차이가 있다***고 해석할 수 있습니다.
- p-값이 0.05보다 크면, 두 그룹 간의 고지혈증 발병 여부에 유의미한 차이가 없다고 해석할 수 있습니다.

In [ ]:
crosstab2

고지혈증_진단_여부_종합    0   1
꾸준히_비음주자              
0              464  61
1              256  56

#### 고혈압

In [ ]:
# 1차~4차 중 한번이라도 고혈압 발병이 있으면 1, 그렇지 않으면 0
filled_merge_df['고혈압_진단_여부_종합'] = (filled_merge_df['고혈압_병력_1차'] == 2) | \
                                 (filled_merge_df['고혈압_병력_2차'] == 2) | \
                                 (filled_merge_df['고혈압_병력_3차'] == 2) | \
                                 (filled_merge_df['고혈압_병력_4차'] == 2)

# 고혈압 발병 여부를 0(발병 안함), 1(발병 있음)으로 변환
filled_merge_df['고혈압_진단_여부_종합'] = filled_merge_df['고혈압_진단_여부_종합'].astype(int)
filled_merge_df['고혈압_진단_여부_종합'].value_counts()

고혈압_진단_여부_종합
0    578
1    259
Name: count, dtype: int64

In [ ]:
import pandas as pd
import scipy.stats as stats

# 꾸준히 비음주자인 사람과 아닌 사람 간의 당뇨병 진단 여부에 대한 교차표 작성
crosstab3 = pd.crosstab(filled_merge_df['꾸준히_비음주자'], filled_merge_df['고지혈증_진단_여부_종합'])

# 카이제곱 검정 수행
chi2, p, dof, expected = stats.chi2_contingency(crosstab3)

# 결과 출력
print(f"카이제곱 통계량: {chi2}")
print(f"p-value: {p}")


카이제곱 통계량: 6.004760843723586
p-value: 0.014267327756523744


- p-값이 0.05보다 작으면, ***꾸준히 비음주자인 사람과 그렇지 않은 사람 간에 고혈압 발병 여부가 통계적으로 유의미한 차이가 있다***고 해석할 수 있습니다.
- p-값이 0.05보다 크면, 두 그룹 간의 고혈압 발병 여부에 유의미한 차이가 없다고 해석할 수 있습니다.

In [ ]:
crosstab3

고지혈증_진단_여부_종합    0   1
꾸준히_비음주자              
0              464  61
1              256  56



---



---



---



---



##가설2) 성인병 발병에 유전적 요인(부모의 병력)이 영향을 미친다.

### 당뇨병

In [ ]:
# 가설 2: 유전적 요인이 영향을 미치는지 검정 (부모의 병력만 사용)

# 1차~4차 중 한번이라도 아버지에게 당뇨병 병력이 있으면 1(발병 있음), 그렇지 않으면 0(발병 없음)
filled_merge_df['아버지_당뇨병_병력_종합'] = (filled_merge_df['부모_당뇨병_병력_아버지_1차'] == 2) | \
                                   (filled_merge_df['부모_당뇨병_병력_아버지_2차'] == 2) | \
                                   (filled_merge_df['부모_당뇨병_병력_아버지_3차'] == 2) | \
                                   (filled_merge_df['부모_당뇨병_병력_아버지_4차'] == 2)

# 아버지 당뇨병 병력을 0(발병 없음), 1(발병 있음)으로 변환
filled_merge_df['아버지_당뇨병_병력_종합'] = filled_merge_df['아버지_당뇨병_병력_종합'].astype(int)

filled_merge_df['아버지_당뇨병_병력_종합'].value_counts()
#

아버지_당뇨병_병력_종합
0    807
1     30
Name: count, dtype: int64

In [ ]:
import pandas as pd
import scipy.stats as stats

# 아버지 당뇨병 병력과 당뇨병 발병 여부에 대한 교차표 작성
crosstab_아버지 = pd.crosstab(filled_merge_df['아버지_당뇨병_병력_종합'], filled_merge_df['당뇨병_진단_여부_종합'])

# 아버지 당뇨병 병력에 대한 카이제곱 검정 수행
chi2_아버지, p_아버지, dof_아버지, expected_아버지 = stats.chi2_contingency(crosstab_아버지)
print(f"아버지 당뇨병 병력 카이제곱 검정 결과: chi2 = {chi2_아버지}, p-value = {p_아버지}")

아버지 당뇨병 병력 카이제곱 검정 결과: chi2 = 1.3813305486818974, p-value = 0.23987463648304513


In [ ]:
# 1차~4차 중 한번이라도 어머니에게 당뇨병 병력이 있으면 1(발병 있음), 그렇지 않으면 0(발병 없음)
filled_merge_df['어머니_당뇨병_병력_종합'] = (filled_merge_df['부모_당뇨병_병력_어머니_1차'] == 2) | \
                                   (filled_merge_df['부모_당뇨병_병력_어머니_2차'] == 2) | \
                                   (filled_merge_df['부모_당뇨병_병력_어머니_3차'] == 2) | \
                                   (filled_merge_df['부모_당뇨병_병력_어머니_4차'] == 2)

# 어머니 당뇨병 병력을 0(발병 없음), 1(발병 있음)으로 변환
filled_merge_df['어머니_당뇨병_병력_종합'] = filled_merge_df['어머니_당뇨병_병력_종합'].astype(int)

filled_merge_df['어머니_당뇨병_병력_종합'].value_counts()

어머니_당뇨병_병력_종합
0    779
1     58
Name: count, dtype: int64

In [ ]:
# 어머니 당뇨병 병력과 당뇨병 발병 여부에 대한 교차표 작성
crosstab_어머니 = pd.crosstab(filled_merge_df['어머니_당뇨병_병력_종합'], filled_merge_df['당뇨병_진단_여부_종합'])

# 어머니 당뇨병 병력에 대한 카이제곱 검정 수행
chi2_어머니, p_어머니, dof_어머니, expected_어머니 = stats.chi2_contingency(crosstab_어머니)
print(f"어머니 당뇨병 병력 카이제곱 검정 결과: chi2 = {chi2_어머니}, p-value = {p_어머니}")

어머니 당뇨병 병력 카이제곱 검정 결과: chi2 = 10.939191856106468, p-value = 0.0009415122713026141


- 여기선 특이하게 아버지, 어머니 따로 나눠서 검정하니까 아버지 병력이랑 당뇨병은 별 관계 없고 , 어머니랑은 관계가 있다고 뜸.

### 고혈압

In [ ]:
# 1차~4차 중 한번이라도 아버지에게 고혈압 병력이 있으면 1(발병 있음), 그렇지 않으면 0(발병 없음)
filled_merge_df['아버지_고혈압_병력_종합'] = (filled_merge_df['부모_고혈압_병력_아버지_1차'] == 2) | \
                                   (filled_merge_df['부모_고혈압_병력_아버지_2차'] == 2) | \
                                   (filled_merge_df['부모_고혈압_병력_아버지_3차'] == 2) | \
                                   (filled_merge_df['부모_고혈압_병력_아버지_4차'] == 2)

# 아버지 고혈압 병력을 0(발병 없음), 1(발병 있음)으로 변환
filled_merge_df['아버지_고혈압_병력_종합'] = filled_merge_df['아버지_고혈압_병력_종합'].astype(int)

filled_merge_df['아버지_고혈압_병력_종합'].value_counts()

아버지_고혈압_병력_종합
0    746
1     91
Name: count, dtype: int64

In [ ]:
# 아버지 고혈압 병력과 고혈압 발병 여부에 대한 교차표 작성
crosstab2_아버지 = pd.crosstab(filled_merge_df['아버지_고혈압_병력_종합'], filled_merge_df['고혈압_진단_여부_종합'])

# 아버지 고혈압 병력에 대한 카이제곱 검정 수행
chi2_아버지, p_아버지, dof_아버지, expected_아버지 = stats.chi2_contingency(crosstab2_아버지)
print(f"아버지 고혈압 병력 카이제곱 검정 결과: chi2 = {chi2_아버지}, p-value = {p_아버지}")

아버지 고혈압 병력 카이제곱 검정 결과: chi2 = 6.170238583169253, p-value = 0.012991708703218835


In [ ]:
# 1차~4차 중 한번이라도 어머니에게 고혈압 병력이 있으면 1(발병 있음), 그렇지 않으면 0(발병 없음)
filled_merge_df['어머니_고혈압_병력_종합'] = (filled_merge_df['부모_고혈압_병력_어머니_1차'] == 2) | \
                                   (filled_merge_df['부모_고혈압_병력_어머니_2차'] == 2) | \
                                   (filled_merge_df['부모_고혈압_병력_어머니_3차'] == 2) | \
                                   (filled_merge_df['부모_고혈압_병력_어머니_4차'] == 2)

# 어머니 고혈압 병력을 0(발병 없음), 1(발병 있음)으로 변환
filled_merge_df['어머니_고혈압_병력_종합'] = filled_merge_df['어머니_고혈압_병력_종합'].astype(int)

filled_merge_df['어머니_고혈압_병력_종합'].value_counts()

어머니_고혈압_병력_종합
0    739
1     98
Name: count, dtype: int64

In [ ]:
# 어머니 고혈압 병력과 고혈압 발병 여부에 대한 교차표 작성
crosstab2_어머니 = pd.crosstab(filled_merge_df['어머니_고혈압_병력_종합'], filled_merge_df['고혈압_진단_여부_종합'])

# 어머니 고혈압 병력에 대한 카이제곱 검정 수행
chi2_어머니, p_어머니, dof_어머니, expected_어머니 = stats.chi2_contingency(crosstab2_어머니)
print(f"어머니 고혈압 병력 카이제곱 검정 결과: chi2 = {chi2_어머니}, p-value = {p_어머니}")

어머니 고혈압 병력 카이제곱 검정 결과: chi2 = 9.388181438636728, p-value = 0.002183887320082561


### 부모 둘다

In [ ]:


# 종속 변수: 당뇨병 발병 여부 (당뇨병_진단_여부_종합)
y = filled_merge_df['당뇨병_진단_여부_종합']

# 독립 변수: 아버지와 어머니의 당뇨병 병력 및 상호작용 항 추가
filled_merge_df['상호작용_항'] = filled_merge_df['아버지_당뇨병_병력_종합'] * filled_merge_df['어머니_당뇨병_병력_종합']

X = filled_merge_df[['아버지_당뇨병_병력_종합', '어머니_당뇨병_병력_종합', '상호작용_항']]

# 상수항 추가 (intercept 추가)
X = sm.add_constant(X)

# 로지스틱 회귀 모델 적합
logit_model = sm.Logit(y, X)
result = logit_model.fit()

# 결과 요약 출력
result.summary()


Optimization terminated successfully.
         Current function value: 0.351099
         Iterations 6


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:           당뇨병_진단_여부_종합   No. Observations:                  837
Model:                          Logit   Df Residuals:                      833
Method:                           MLE   Df Model:                            3
Date:                Mon, 23 Sep 2024   Pseudo R-squ.:                 0.02107
Time:                        11:33:07   Log-Likelihood:                -293.87
converged:                       True   LL-Null:                       -300.19
Covariance Type:            nonrobust   LLR p-value:                  0.005459
=================================================================================
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
const            -2.1565      0.120    -18.031      0.000      -2.391      -1.922
아버지_당뇨병_병력_종합     0.4073      0.555      0.734      0.463      -0.680       1.495
어머니_당뇨병_병력_종합     0.9838      0.339      2.901      0.004       0.319       1.649
상호작용_항            1.4585      1.381      1.056      0.291      -1.249       4.166
=================================================================================
"""

- Pseudo R-squared (0.02107):

이 값은 회귀 모델이 종속 변수를 얼마나 잘 설명하는지를 나타내는 지표입니다. 일반적인 R-squared와 달리, Pseudo R-squared는 로지스틱 회귀에서 사용되며, 모델의 설명력을 나타냅니다.
0.02107은 모델이 2.1% 정도의 변동성을 설명한다는 의미로, 모델의 설명력이 낮습니다. 즉, 부모의 병력만으로 당뇨병 발병을 예측하는 데 한계가 있음을 보여줍니다.
- LLR p-value (0.005459):

모델 전체의 유의성을 평가하는 지표. p-value = 0.0055로, 0.05보다 작으므로 모델 전체가 통계적으로 유의미하다는 것을 의미 => 즉, 독립 변수들이 종속 변수에 유의미한 영향을 미친다고 해석

- 각 변수 해석:
1. 상수항(const): -2.1565 (p-value < 0.001)
상수항은 독립 변수 값이 0일 때 당뇨병 발병 여부의 로그 오즈(확률).
이 값이 음수이므로, 기본적으로 당뇨병 발병 확률은 낮다는 것을 의미
2. 아버지_당뇨병_병력_종합 (coef: 0.4073, p-value: 0.463)
회귀 계수(coef): 아버지의 당뇨병 병력이 있을 때 당뇨병 발병 확률이 증가하는 방향으로 영향을 미치지만, p-value = 0.463로 유의미하지 않다고 해석됩니다.     **즉, 아버지의 당뇨병 병력은 당뇨병 발병 여부에 통계적으로 유의미한 영향을 미치지 않는다**고 할 수 있음.
3. 어머니_당뇨병_병력_종합 (coef: 0.9838, p-value: 0.004)
회귀 계수(coef): 어머니의 당뇨병 병력이 있는 경우, 당뇨병 발병 확률이 증가.
p-value = 0.004로, 통계적으로 유의미한 결과입니다. 즉, **어머니의 당뇨병 병력은 당뇨병 발병에 유의미한 영향을 미친다**고 해석
**회귀 계수(0.9838)**는 로그 오즈 비로 해석되며, 이는 당뇨병 발병 확률이 어머니의 병력이 있을 때 증가한다는 의미.
4. 상호작용_항 (coef: 1.4585, p-value: 0.291)
회귀 계수(coef): 아버지와 어머니의 당뇨병 병력이 모두 있을 때, 당뇨병 발병 확률이 증가하는 방향으로 영향을 미치지만, p-value = 0.291로 통계적으로 유의미하지 않다고 해석됩니다. 즉, **아버지와 어머니의 병력 간의 상호작용 효과는 유의미하지 않다**고 할 수 있음.


결론:
- 어머니의 당뇨병 병력은 당뇨병 발병에 유의미한 영향을 미치며, 어머니의 병력이 있을 때 당뇨병 발병 확률이 높아진다.
- 아버지의 당뇨병 병력은 당뇨병 발병에 통계적으로 유의미한 영향을 미치지 않는다.
- 상호작용 항 역시 통계적으로 유의미하지 않으므로, 아버지와 어머니의 당뇨병 병력이 동시에 있을 때 특별히 더 높은 당뇨병 발병 확률을 보이는 상호작용 효과는 없다.
- 이 분석 결과에 따라, 어머니의 병력이 당뇨병 발병에 더 중요한 영향을 미치는 것으로 보임.

In [ ]:
# 종속 변수: 고혈압 발병 여부 (고혈압_진단_여부_종합)
y = filled_merge_df['고혈압_진단_여부_종합']

# 독립 변수: 아버지와 어머니의 고혈압 병력 및 상호작용 항 추가
filled_merge_df['상호작용_항'] = filled_merge_df['아버지_고혈압_병력_종합'] * filled_merge_df['어머니_고혈압_병력_종합']

X = filled_merge_df[['아버지_고혈압_병력_종합', '어머니_고혈압_병력_종합', '상호작용_항']]

# 상수항 추가 (intercept 추가)
X = sm.add_constant(X)

# 로지스틱 회귀 모델 적합
logit_model = sm.Logit(y, X)
result = logit_model.fit()

# 결과 요약 출력
result.summary()

Optimization terminated successfully.
         Current function value: 0.610089
         Iterations 5


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:           고혈압_진단_여부_종합   No. Observations:                  837
Model:                          Logit   Df Residuals:                      833
Method:                           MLE   Df Model:                            3
Date:                Mon, 23 Sep 2024   Pseudo R-squ.:                 0.01384
Time:                        11:33:08   Log-Likelihood:                -510.64
converged:                       True   LL-Null:                       -517.81
Covariance Type:            nonrobust   LLR p-value:                  0.002486
=================================================================================
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
const            -0.9477      0.086    -11.031      0.000      -1.116      -0.779
아버지_고혈압_병력_종합     0.5800      0.265      2.191      0.028       0.061       1.099
어머니_고혈압_병력_종합     0.6999      0.251      2.788      0.005       0.208       1.192
상호작용_항           -0.4122      0.535     -0.771      0.441      -1.460       0.636
=================================================================================
"""

- 어머니의 고혈압 병력은 고혈압 발병에 유의미한 영향을 미치며, 어머니의 병력이 있을 때 고혈압 발병 확률이 높아진다.
- 아버지의 고혈압 병력은 고혈압 발병에 통계적으로 유의미한 영향을 미친다.
- 상호작용 항은 통계적으로 유의미하지 않으므로, 아버지와 어머니의 고혈압 병력이 동시에 있을 때 특별히 더 높은 고혈압 발병 확률을 보이는 상호작용 효과는 없다.

## 가설3) 음주, 가족력 둘 중에 성인병 발생에 더 큰 영향을 미치는 변수는?

### 랜덤포레스트(당뇨)

In [ ]:
from sklearn.ensemble import RandomForestClassifier

# 독립 변수 설정
X = filled_merge_df[['꾸준히_비음주자', '아버지_당뇨병_병력_종합', '어머니_당뇨병_병력_종합']]

# 종속 변수 설정: 당뇨병 발병 여부 (0=발병 안함, 1=발병 있음)
y = filled_merge_df['당뇨병_진단_여부_종합']

# 랜덤 포레스트 분류 모델 설정
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)

# 모델 학습
rf_model.fit(X, y)

# 변수 중요도 출력
feature_importances = pd.Series(rf_model.feature_importances_, index=X.columns)
print(feature_importances)

# 중요도 순으로 정렬
feature_importances = feature_importances.sort_values(ascending=False)
feature_importances


꾸준히_비음주자         0.289247
아버지_당뇨병_병력_종합    0.207439
어머니_당뇨병_병력_종합    0.503314
dtype: float64


어머니_당뇨병_병력_종합    0.503314
꾸준히_비음주자         0.289247
아버지_당뇨병_병력_종합    0.207439
dtype: float64

### 랜덤포레스트(고혈압)

In [ ]:
from sklearn.ensemble import RandomForestClassifier

# 독립 변수 설정
X1 = filled_merge_df[['꾸준히_비음주자', '아버지_고혈압_병력_종합', '어머니_고혈압_병력_종합']]

# 종속 변수 설정: 고혈압 발병 여부 (0=발병 안함, 1=발병 있음)
y1 = filled_merge_df['고혈압_진단_여부_종합']

# 랜덤 포레스트 분류 모델 설정
rf_model1 = RandomForestClassifier(n_estimators=100, random_state=42)

# 모델 학습
rf_model1.fit(X1, y1)

# 변수 중요도 출력
feature_importances2 = pd.Series(rf_model1.feature_importances_, index=X1.columns)
print(feature_importances2)

# 중요도 순으로 정렬
feature_importances2 = feature_importances2.sort_values(ascending=False)
feature_importances2

꾸준히_비음주자         0.26859
아버지_고혈압_병력_종합    0.29280
어머니_고혈압_병력_종합    0.43861
dtype: float64


어머니_고혈압_병력_종합    0.43861
아버지_고혈압_병력_종합    0.29280
꾸준히_비음주자         0.26859
dtype: float64

### 로지스틱 회귀(당뇨)

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

# 독립 변수와 종속 변수 설정
X = filled_merge_df[['꾸준히_비음주자', '아버지_당뇨병_병력_종합', '어머니_당뇨병_병력_종합']]
y = filled_merge_df['당뇨병_진단_여부_종합']

# 변수 표준화
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# 로지스틱 회귀 모델 적합
log_reg = LogisticRegression()
log_reg.fit(X_scaled, y)

# 회귀 계수 확인
coefficients = log_reg.coef_[0]
features = X.columns

coefficients_df = pd.DataFrame({'Feature': features, 'Coefficient': coefficients})
coefficients_df

Feature  Coefficient
0       꾸준히_비음주자     0.290010
1  아버지_당뇨병_병력_종합     0.132530
2  어머니_당뇨병_병력_종합     0.279504

### 로지스틱 회귀 (고혈압)

In [ ]:
# 독립 변수와 종속 변수 설정
X2 = filled_merge_df[['꾸준히_비음주자', '아버지_고혈압_병력_종합', '어머니_고혈압_병력_종합']]
y2 = filled_merge_df['당뇨병_진단_여부_종합']

# 변수 표준화
scaler = StandardScaler()
X2_scaled = scaler.fit_transform(X2)

# 로지스틱 회귀 모델 적합
log_reg = LogisticRegression()
log_reg.fit(X2_scaled, y2)

# 회귀 계수 확인
coefficients2 = log_reg.coef_[0]
features2 = X2.columns

coefficients_df = pd.DataFrame({'Feature': features2, 'Coefficient': coefficients2})
coefficients_df

Feature  Coefficient
0       꾸준히_비음주자     0.280919
1  아버지_고혈압_병력_종합     0.080522
2  어머니_고혈압_병력_종합     0.088998

# 관련 자료

제 2형 당뇨병은 유전적 요인에 의해서
약 50% 정도 영향을 받는 것으로
알려져 있습니다

부모가 모두 제 2형 당뇨병인 경우
자녀가 제 2형 당뇨병이 발병할 소지는
30% 정도이고,
부모 중 한 사람만 제 2형 당뇨병인 경우
자녀가 제 2형 당뇨병이 발병할 소지는
15% 정도입니다.

하지만 가족 중에 제 2형 당뇨병 환자가
있다고 해서 반드시 제 2형 당뇨병이
발병되는 것은 아니며,

반대로 가족 중에
제 2형 당뇨병 환자가 없다고 해서
제 2형 당뇨병 발병 위험이
전혀 없는 것은 아닙니다.

그만큼 제 2형 당뇨병 발병에
환경적 요인도 많은 영향을 주고 있습니다.

[한양대학교 병원 자료](https://seoul.hyumc.com/seoul/healthInfo/healthLife.do?action=view&bbsId=healthLife&nttSeq=12555)